server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionInstaller.java

```diff
@@ -0,0 +1,146 @@
+    MutableEditionManagementState editionManagementState) {
+    this.editionPluginDownloader = editionDownloader;
+    this.editionPluginUninstaller = editionPluginUninstaller;
+    this.pluginRepository = pluginRepository;
+    this.executor = executor;
+    this.updateCenterMatrixFactory = updateCenterMatrixFactory;
+    this.editionManagementState = editionManagementState;
+  }
+
+  /**
+   * Refreshes the update center, and submits in a executor a task to download all the needed plugins (asynchronously).
+   * If the update center is disabled or if we are offline, the task is not submitted and false is returned.
+   * In all case
+   *
+   * @throws IllegalStateException if an installation is already in progress
+   */
+  public void install(License newLicense) {
+    if (semaphore.tryAcquire()) {
+      try {
+        Optional<UpdateCenter> updateCenter = updateCenterMatrixFactory.getUpdateCenter(true);
+        if (!updateCenter.isPresent()) {
+          editionManagementState.startManualInstall(newLicense);
+          return;
+        }
+        editionManagementState.startAutomaticInstall(newLicense);
+        executor.execute(() -> asyncInstall(newLicense, updateCenter.get()));
+      } catch (RuntimeException e) {
+        semaphore.release();
+        throw e;
+      }
+    } else {
+      throw new IllegalStateException("Another installation of an edition is already running");
+    }
+  }
+
+  public void uninstall() {
+    Map<String, PluginInfo> pluginInfosByKeys = pluginRepository.getPluginInfosByKeys();
+    Set<String> pluginsToRemove = pluginsToRemove(Collections.emptySet(), pluginInfosByKeys.values());
+    uninstallPlugins(pluginsToRemove);
+  }
+
+  /**
+   * Check if the update center is disabled or unreachable. It uses the cached status (it doesn't refresh),
+   * to be a cost-free check.
+   */
+  public boolean isOffline() {
+    return !updateCenterMatrixFactory.getUpdateCenter(false).isPresent();
+  }
+
+  public boolean requiresInstallationChange(Set<String> editionPluginKeys) {
+    Map<String, PluginInfo> pluginInfosByKeys = pluginRepository.getPluginInfosByKeys();
+
+    return !pluginsToInstall(editionPluginKeys, pluginInfosByKeys.keySet()).isEmpty()
+      || !pluginsToRemove(editionPluginKeys, pluginInfosByKeys.values()).isEmpty();
+  }
+
+  private void asyncInstall(License newLicense, UpdateCenter updateCenter) {
+    try {
+      Set<String> editionPluginKeys = newLicense.getPluginKeys();
+      Map<String, PluginInfo> pluginInfosByKeys = pluginRepository.getPluginInfosByKeys();
+      Set<String> pluginsToRemove = pluginsToRemove(editionPluginKeys, pluginInfosByKeys.values());
+      Set<String> pluginsToInstall = pluginsToInstall(editionPluginKeys, pluginInfosByKeys.keySet());
+
+      editionPluginDownloader.downloadEditionPlugins(pluginsToInstall, updateCenter);
+      uninstallPlugins(pluginsToRemove);
+      editionManagementState.automaticInstallReady();
+    } catch (Throwable t) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1181)